File: 1621-number-of-subsequences-that-satisfy-the-given-sum-condition/1621-number-of-subsequences-that-satisfy-the-given-sum-condition.java
class Solution {
    private static final long MOD = 1_000_000_007L;

    public int numSubseq(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;
        int[] pow2 = new int[n];
        pow2[0] = 1;

        for (int i = 1; i < n; i++) {
            pow2[i] = (int)((pow2[i - 1] * 2L) % MOD);  // cast after long calc
        }

        int left = 0, right = n - 1;
        int res = 0;

        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                res = (int)((res + pow2[right - left]) % MOD);  // cast here too
                left++;
            } else {
                right--;
            }
        }

        return res;
    }
}

File: 1621-number-of-subsequences-that-satisfy-the-given-sum-condition/1621-number-of-subsequences-that-satisfy-the-given-sum-condition.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function(nums, target) {
    const MOD = 1_000_000_007;
    nums.sort((a, b) => a - b);
    const n = nums.length;

    // Precompute powers of 2 modulo MOD
    const pow2 = Array(n).fill(1);
    for (let i = 1; i < n; i++) {
        pow2[i] = (pow2[i - 1] * 2) % MOD;
    }

    let res = 0;
    let left = 0, right = n - 1;

    while (left <= right) {
        if (nums[left] + nums[right] <= target) {
            res = (res + pow2[right - left]) % MOD;
            left++;
        } else {
            right--;
        }
    }

    return res;
};